File: nim/src/autd3/simulator.nim
# File: simulator.nim
# Project: autd3
# Created Date: 08/08/2022
# Author: Shun Suzuki
# -----
# Last Modified: 28/04/2023
# Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
# -----
# Copyright (c) 2022 Shun Suzuki. All rights reserved.
#



import native_methods/autd3capi_link_simulator
import link

type Simulator* = object of RootObj
  builder: pointer

proc initSimulator*(): Simulator =
  AUTDLinkSimulator(result.builder.addr)

proc logLevel*(cnt: var Simulator, level: int32): var Simulator =
  AUTDLinkSimulatorLogLevel(cnt.builder, level)
  result = cnt
  
proc logFunc*(cnt: var Simulator, logOut: LogOut, logFlush: LogFlush): var Simulator =
  let pLogOut = rawProc(logOut)
  let pLogFlush = rawProc(logFlush)
  AUTDLinkSimulatorLogFunc(cnt.builder, pLogOut, pLogFlush)
  result = cnt

proc timeout*(cnt: var Simulator, timeout: uint64): var Simulator =
  AUTDLinkSimulatorTimeout(cnt.builder, timeout)
  result = cnt

func build*(cnt: Simulator): Link =
  AUTDLinkSimulatorBuild(result.p.addr, cnt.builder)

File: matlab/autd3/Simulator.m
%{
%File: Simulator.m
%Project: autd3
%Created Date: 10/10/2022
%Author: Shun Suzuki
%-----
%Last Modified: 28/04/2023
%Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
%-----
%Copyright (c) 2022 Shun Suzuki. All rights reserved.
%
%}

classdef Simulator < handle

    properties
        ptr
        builder_
    end

    methods

        function obj = Simulator()
            obj.ptr = libpointer('voidPtr', 0);
            obj.builder_ = libpointer('voidPtr', 0);
            pp = libpointer('voidPtrPtr', obj.builder_);
            calllib('autd3capi_link_simulator', 'AUTDLinkSimulator', pp);
        end

        function res = timeout(obj, t)
            calllib('autd3capi_link_simulator', 'AUTDLinkSimulatorTimeout', obj.builder_, t);
            res = obj;
        end
        
        function res = build(obj)
            pp = libpointer('voidPtrPtr', obj.ptr);
            calllib('autd3capi_link_simulator', 'AUTDLinkSimulatorBuild', pp, obj.builder_);
            res = obj;
        end

    end

end 
